### ISSUE TITLE
Update concept

### ISSUE THREAD

**Issue by probonopd** [simoniz0r/spm#6]
Not sure I entirely understand your update concept here, but with AppImages the authoritative information about where an AppImage should get updates from is _inside the AppImage itself_ as per the AppImageSpec:

https://github.com/AppImage/AppImageSpec/blob/master/draft.md#update-information

Trying to load information about the availability of potentially available updates from any other location than the update information embedded inside the AppImage itself is futile. Please do not do it, as it would be against the AppImage philosophy which mandates that update information has to travel inside the (signed) AppImage itself, not via a side-channel (which can easily be, and will in practice always be, outdated). Also, it would not work with [AppImageUpdate](https://github.com/AppImage/AppImageUpdate) binary delta (differential) updates.

https://github.com/simoniz0r/spm/blob/e432ce659ddf429fbe9ee277fc3a4e0c47a3a0ef/appimgfunctions.sh#L213

**Comment by simoniz0r** [comment]
spm uses github's API to check the `updated_at` key to see when the latest AppImage has been released on each repo.  I've tested this pretty extensively, and it works very well.  As soon as a new AppImage is added (even if it's added to an existing tag), the `updated_at` key changes, and spm marks that AppImage for upgrading.

I tried AppImageUpdate on a few AppImages, but it only worked with a couple of them, and the way that it works was not really what I want at all.  It failed to update most AppImages that I tried (I assume they do not contain the necessary data).  For the couple of AppImages that it did work on, it just downloads the latest AppImage regardless of whether the current one is up to date or not, and then it does not replace the existing AppImage.  It just dumps the latest AppImage in the same directory as the AppImage that's being upgraded.

The method that I'm using with GIthub's API to check the `updated_at` key seems to be much more reliable and predictable.

**Comment by probonopd** [comment]
> I tried AppImageUpdate on a few AppImages, but it only worked with a couple of them, and the way that it works was not really what I want at all. It failed to update most AppImages that I tried (I assume they do not contain the necessary data).

Yes, this is because most AppImages do not yet contain update information. This should change if we make tools such as spm actually _use_ it. Working around by just ignoring the update information will only worsen the situation because application authors will think "updates work with it anyway". So at the very least we should give _preferece_ to update information inside the AppImage, and give a clear _warning_ when we _resort to_ scraping URLs from GitHub or other locations.

>  As far as I can tell, there is no way for me to just run AppImageUpdate to check if there is a new AppImage without the AppImage being downloaded.

Why would that even be useful? The way I see it, a user runs AppImageUpdate on an AppImage to see if a newer version is available, and if it is, "update" the AppImage (i.e., download a binary delta to the new one, while leaving the old one intact).
Would having an option to run AppImageUpdate on an entire _directory_ fulfill your needs? It would essentially run AppImageUpdate on every AppImage that a user has.

> The method that I'm using with GIthub's API to check the updated_at

Perhaps we could work that into AppImageUpdate, and use it __if__ the AppImage author has specified GitHub in the [update information](https://github.com/AppImage/AppImageSpec/blob/master/draft.md#update-information).

> As a side note, I think that AppImageUpdate should be an optional tool. If distros were to decide to add AppImages to their repos somehow, using AppImageUpdate on those AppImages would more than likely cause the distro to not know what version the AppImage is.

I don't think distros _should_ add AppImages to their repos, at least not in the traditional sense of package management. Have a look at the [Sparkle framework](https://sparkle-project.org/) for macOS, possibly we can learn from that, as it is more in line with the decentralized vision of AppImage than the traditional disto model.

Don't get me wrong, I really like what you are doing, but let's try to keep it close to the vision of AppImage, not having to install stuff, not needing root, being as decentral and as close to upstream as possible.

**Comment by simoniz0r** [comment]
> Would having an option to run AppImageUpdate on an entire directory fulfill your needs? It would essentially run AppImageUpdate on every AppImage that a user has.

Yeah, as long as it can be done in an efficient, is reliable, and has options to run from the command line without a GUI.  

Honestly, it seems like it puts a little extra burden on people making AppImages to maintain a server or consistent redirecting link that can easily be taken care of if they just host things on places like Github with easy to use APIs.  I like the idea of working getting info from github's API into AppImageUpdate.

> I don't think distros should add AppImages to their repos, at least not in the traditional sense of package management.

I don't really see why not.  If it's easier on the developers to make an AppImage than to create a package for that distro, AppImages could be a nice alternative to things like flatpaks and snaps that are currently being used in some distros package managers in some shape or form.  AppImages don't require any other packages to be installed to work, unlike flatpaks and snaps which need their package managers to be installed to install them.

I don't think AppImages should be used for every package out there, but for very simple applications (such as something like neofetch which is basically a bash script), AppImages could be a nice way to distribute them without having to build packages for each distrio.

**Comment by probonopd** [comment]
I agree that AppImage is useful for many packages, but what value would be added by distributions distributing AppImages that the user could get just as well directly from the original author?

I think that distributions can help _authors to make AppImages_ (via systems such as the openSUSE Build Service), but have doubts that distribution package mangers should play a role in distributing the resulting AppImages.

**Comment by probonopd** [comment]
> Honestly, it seems like it puts a little extra burden on people making AppImages to maintain a server or consistent redirecting link

That's true. But you gain binary delta updates, something that is really useful in many scenarios. Think of large applications or games (a few GB are not uncommon). Do you really want to re-download the whole thing each time a minor update comes out?

**Comment by simoniz0r** [comment]
> I agree that AppImage is useful for many packages, but what value would be added by distributions distributing AppImages that the user could get just as well directly from the original author?

It seems to me that, for the most part, people mainly stick to packages that are available in their repos.  Having AppImages in distros repos would allow the developers to build one very easy to create package that would just work everywhere.  This would allow users more access to software that they would otherwise more than likely not be aware of at all.

> That's true. But you gain binary delta updates, something that is really useful in many scenarios. Think of large applications or games (a few GB are not uncommon). Do you really want to re-download the whole thing each time a minor update comes out?

Honestly, when it comes to large programs, if a native package is available, I'm going to use that.  I don't really like the idea of using universal packages for large software with a huge amount of dependencies unless those dependencies are very hard to find.  

Most of the time, the universal package ends up being quite large, and it contains many dependencies that I already have installed on my system.  This causes the universal package to be quite larger than just installing a native package.

I think AppImages are great for simple software with relatively few dependencies.  I also think they're great for software such as KDE software where the user would have to install a boat load of dependencies to use that software if they are not on KDE already.

**Comment by probonopd** [comment]
> Having AppImages in distros repos would allow the developers to build one very easy to create package that would just work everywhere.

If you talk to distros, they will probably tell you that they would not accept such kinds of "package" (really: a bundle-in-an-AppImage) since they would violate their [packaging policies](https://www.debian.org/doc/debian-policy/) in substancial ways - it's two different philosophies, each with its advantages and downsides. Combining the downsides of traditional package management with the downsides of a bundle approach (like AppImage, e.g., larger download sizes) doesn't seem very attractive to me...

**Comment by simoniz0r** [comment]
If that's the case, then I'm kind of confused as to why many distros have taken/are taking steps to add flatpaks and snaps to their software centers.

I think that AppImages would be great for an extremely simple application with either very few dependencies or no dependencies other than that would be on the base system.  Programs like neofetch, inxi, etc that are pretty much just bash scripts with some extra files that they need to work run very well in an AppImage.

As I stated before, for larger programs with many dependencies, I pretty much always prefer a native package as opposed to universal packages, but AppImages seem to be quite a bit leaner than flatpaks or snaps.  The last time I tried to install a flatpak, it wanted to install over 200MBs worth of files for something that had a 4MB install size in my repos.  The program that I was trying to install was a very simple program (as simple if not more simple than what is being bundled in AppImages currently), but it was still over double the install size of any AppImage that I've seen.

**Comment by probonopd** [comment]
> If that's the case, then I'm kind of confused as to why many distros have taken/are taking steps to add flatpaks and snaps to their software centers.

Flatpak is much closer to a tarditional package manager in that it allows for dependencies and uses package manager tools to manage the applications. To a lesser extent, this is also true for snappy. Both come from the distro world, unlike AppImage, which takes inspiration from macOS.

> As I stated before, for larger programs with many dependencies, I pretty much always prefer a native package as opposed to universal packages

Especially for such applications I like AppImage, because I never want to touch anything in my base system, no matter how many (and how bleeding-edge) applications I am running. I don't want my installation of app X result in half of the base system being updated. Enterprises may think in a similar way, or not even give their employees root rights and the ability to change anything in the system.

>  The last time I tried to install a flatpak, it wanted to install over 200MBs worth of files for something that had a 4MB install size in my repos.

That is because Flatpak has the concept of dependencies on runtimes, and runtimes can be rather large thing such as "Gnome". Even if you are just interested in an application that uses a tiny fraction of Gnome, it will pull in the whole Gnome stack. In contrast, well-made AppImages bundle just the dependencies that are really needed for the particular application at hand. For example, the `linuxdeployqt` tool allows authors to make AppImages from Qt apps by bundling just the very _portions of_ Qt that the _specific_ application needs. Also, AppImages are compressed at all time (since they are _not_ installed), which adds to their on-disk efficiency. In fact, LibreOffice in an AppImage is _smaller_ than the same LibreOffice installed from debs or rpms.

But hey, before this drifts into a general conversation about different bundling formats, let's come back to the topic of the issue: Promoting the AppImage way of doing binary delta updates, using update information embedded inside the AppImages themselves. The optional `appimaged` daemon will start making use of this soon, and we will also push it on AppImageHub. So it would be nice if spm would at least not _ignore_ it.

**Comment by simoniz0r** [comment]
> But hey, before this drifts into a general conversation about different bundling formats, let's come back to the topic of the issue: Promoting the AppImage way of doing binary delta updates, using update information embedded inside the AppImages themselves. The optional appimaged daemon will start making use of this soon, and we will also push it on AppImageHub. So it would be nice if spm would at least not ignore it.

I'm still not sure that I'm entirely on board with the concept of AppImageUpdate.  It adds an extra barrier of entry to creating AppImages that doesn't really need to be a thing; as long as links are provided that do not directly link to the current AppImage version, it's pretty easy for a package manager to check those links for new versions.

I also feel like for the majority of AppImages, it would probably just be faster to download the entire AppImage (takes like 30 seconds on a good connection) than using AppImageUpdate to only update parts of the AppImage.

The fact that users can break its functionality easily in a few different ways makes me even more hesitant to use it.  In my opinion, you shouldn't be saying `Don't do X Y or Z with your AppImage because it breaks them`; you should be figuring out ways to prevent X Y or Z from breaking the AppImage because people are probably going to end up doing it no matter how many times you tell them not to.  In my opinion, it's always best to try and be as idiot proof as possible rather than having warning signs all over the place that can easily be missed or ignored.

I haven't been able to get appimaged to run at all due to lib issues, so that kinda takes away my ability to try and integrate that with spm at all.  Both of these tools are good in concept, but in real world use, I'm not finding them to be reliable or efficient enough to consider using in my package manager at this point.

If I were to try using them right now, they would cause far more issues than they would solve.  It would end up being a nightmare for me to try and support something that I can't even get to work reliably...

**Comment by simoniz0r** [comment]
If I could check for updates, get the link to the latest version, and then feed that link into AppImageUpdate, I think that would work out well for me.  If AppImageUpdate acted like less of a package manager itself and was just a tool to be able to apply updates to AppImages efficiently, it would fit my needs a lot better.

spm finds that AppImage X needs an upgrade, spm gets the link for AppImage X, spm runs AppImageUpdate with something like `appimageupdate /usr/local/bin/X https://github.com/someuser/X/releases/download/VERSION/X-VERSION-x86_64.AppImage`, AppImageUpdate does it's thing, and spm marks the package as being upgraded.

Something along those lines would make me consider using AppImageUpdate as it would take the burden off of the developers to maintain working links, and it would allow me to find links to the latest versions in a way that I know works well.

**Comment by probonopd** [comment]
AppImageUpdate (so far) operates from the assumption that no one knows what the "latest" version is, execpt by checking the update information embedded inside the AppImage where the author states where the latest version is.

To do what you suggest, authors would need to have another way to tell spm where the latest version is (and not all projects use GitHub Releases, including very large ones such as KDE). Possibly AppStream could be used by this purpose, but its adoption is also far less than 100%.

**Comment by simoniz0r** [comment]
You wouldn't necessarily have to change the way that AppImageUpdate works now.  You could just add in the ability for the user to be able to specify the link to the latest AppImage version with an argument.

**Comment by probonopd** [comment]
And what should happen then?

**Comment by simoniz0r** [comment]
AppImageUpdate would replace the link that you provide it with the one that the user provides, and then it would go about upgrading the AppImage as it normally would, except for have command line output rather than calling a GUI.

Perhaps if you just run AppImageUpdate without any arguments, the user could be given the option to select an AppImage and input a link if one isn't found.

This would make AppImageUpdate much more versatile than it currently is.  Users could supply their own links if the included link isn't working or if one wasn't included.

**Comment by probonopd** [comment]
>  Users could supply their own links if the included link isn't working or if one wasn't included.

Let's have the burden to be with the AppImage author, not with the user. The AppImage is supposed to know how to "update itself" without the user needing to do anything. It's true that most AppImages don't do this today, but hey, mine do - let's push other providers of AppImages to do the same :-)

**Comment by simoniz0r** [comment]
I'm not suggesting that you take the burden off of the Author if they *want* to provide a link.

I'm suggesting that you give users an option to upgrade AppImages using AppImageUpdate if those AppImages either do not provide a link or their links are no longer working.

**Comment by probonopd** [comment]
PR welcome. Extremely easy to do, as the underlying zsnyc tool normally exactly expects a link to the zsync file as an argument. I see no use in it, since the user doesn't know the zsnyc link.

**Comment by simoniz0r** [comment]
Maybe I'll look into it, but the way I'm upgrading AppImages is working fine for spm's uses.  If you'd like spm to utilize AppImageUpdate, it would be most efficient for spm to be able to interact with AppImageUpdate in the ways that I've suggested above.  It would even have an extra use case as I also suggested above with users being able to use AppImageUpdate in the same way.

I just want spm to provide an efficient, consistent experience when checking for updates and upgrading AppImages, but I'm not really able to do that when AppImageUpdate acts much differently than any other package type that spm does or would handle.
